### DISCUSSION TITLE
SocketIO 3 -> Server API -> get socket objects by ID?

### DISCUSSION THREAD

**Discussion by tgbv** [socketio/socket.io#3804] [Q&A] [answered]
According to docs I can do:
```
await io.of('/').allSockets()
```
....and receive all connected socket IDs.

But how to retrieve the actual socket objects based on those IDs? 🤔

**Comment by darrachequesne** [comment, marked as answer]
You can use the [`namespace#sockets`](https://socket.io/docs/v3/server-api/#namespace-sockets) attribute:

```js
for (const [id, socket] of io.of("/").sockets) {
  console.log(id, socket.handshake);
}
```

Please note that this Map contains only the socket instances on the current Socket.IO server, while the `allSockets()` method returns all socket IDs across multiple Socket.IO servers.

**Reply by tgbv** [reply]
Thank you for reply!

>  contains only the socket instances on the current Socket.IO server
> [ ... ]
> across multiple Socket.IO servers

By "socketio server" you mean socket namespace?

**Reply by darrachequesne** [reply]
In the following example:

![untitled](https://user-images.githubusercontent.com/13031701/108042389-0213c500-7040-11eb-8b96-22adc27509db.png)

Running:

```js
for (const [id, socket] of io.of("/").sockets) {
  console.log(id, socket.handshake);
}
```

on Server A will loop on 2 Socket instances (Client 1 & 2).

But:

```js
const ids = await io.of("/").allSockets();
```

will return 3 Socket IDs (provided that you use the Redis adapter).

Documentation: https://socket.io/docs/v3/using-multiple-nodes/

This is also true for other namespaces:

```js
// loop on the "local" socket instances
for (const [id, socket] of io.of("/admin").sockets) {
  console.log(id, socket.handshake);
}
// fetch all the IDs across Socket.IO servers
const ids = await io.of("/admin").allSockets();
```

Documentation: https://socket.io/docs/v3/namespaces/

**Reply by tgbv** [reply]
Understood! Thank you
